Griddly.Mvc/GriddlyParameterAttribute.cs

```diff
@@ -22,264 +13,251 @@
+                    } 
+                }
+
+                if (IsChildAction(filterContext))
+                {
+                    string[] querystringKeys = GetQueryStringKeys(request);
 
-                    if (IsChildAction(filterContext))
+                    // if a param came in on querystring, skip the defaults. cookie already does its own skip.
+                    if (!context.IsDefaultSkipped && actionArguments.Any(x => x.Value != null && querystringKeys.Contains(x.Key)))
                     {
-                        string[] parentKeys = GetQueryStringKeys(request);
+                        context.IsDefaultSkipped = true;
+                        context.IsDeepLink = true;
+                    }
 
-                        // if a param came in on querystring, skip the defaults. cookie already does its own skip.
-                        if (!context.IsDefaultSkipped && args.Any(x => x.Value != null && parentKeys.Contains(x.Key)))
+                    foreach (var param in actionArguments.ToList())
+                    {
+                        if (param.Value != null && (param.Value.GetType().IsValueType || typeof(IEnumerable).IsAssignableFrom(param.Value.GetType())))
                         {
-                            context.IsDefaultSkipped = true;
-                            context.IsDeepLink = true;
-                        }
+                            bool isParamSet = context.CookieData?.Values?.ContainsKey(param.Key) == true || querystringKeys.Contains(param.Key) || filterContext.RouteData.Values.ContainsKey(param.Key);
 
-                        foreach (var param in args.ToList())
-                        {
-                            if (param.Value != null && (param.Value.GetType().IsValueType || typeof(IEnumerable).IsAssignableFrom(param.Value.GetType())))
+                            if (isParamSet)
                             {
-                                bool isParamSet = context.CookieData?.Values?.ContainsKey(param.Key) == true || parentKeys.Contains(param.Key) || filterContext.RouteData.Values.ContainsKey(param.Key);
-
-                                // if we're skipping defaults but this didn't come from cookie or querystring, it must've come from a
-                                // parameter default in code... nuke it.
-                                if ((context.IsDefaultSkipped && !DefaultIgnoreSkipped) && !isParamSet)
-                                    args[param.Key] = null;
-                                else
-                                {
-                                    if (!(context.IsDefaultSkipped && !DefaultIgnoreSkipped) || !isParamSet)
-                                        context.Defaults[param.Key] = param.Value;
-
-                                    context.Parameters[param.Key] = param.Value;
-                                }
+                                context.Parameters[param.Key] = param.Value;
```

Comment:
Yeah, it took me awhile to wrap my head around it. Essentially, I think this was the lifecycle of the request and filters in the old code:

1. Parent action: SetGriddlyDefault populates `context.Defaults` always
2. Child action:
  1. SetGriddlyDefault populates `context.Parameters` always (because `DefaultIgnoreSkipped` was initialized as true)
  2. If querystring or cookie exists, value provider would reset BOTH `context.Parameters` and `context.Defaults` based on querystring/cookie values (it's a bit convoluted but that's essentially what was happening).

The new request and filter lifecycle looks like this:

1. Parent action: SetGriddlyDefault populates `context.Defaults` always
2. Child action:
  1. SetGriddlyDefault populates `context.Parameters` only if `DefaultIgnoreSkipped` has been set to true, or `context.IsDefaultSkipped` is false (e.g. no cookie or querystring)
  2. If querystring or cookie exists, value provider will reset only `context.Parameters` and leave the defaults untouched.

I added an extra property on the context `IsDefaultReplaced` to enable implementors to do dynamic replacement of default filters (in my use case, I want to load saved filters from the database as the "default" per user).

The result is that the applied filter (`context.Parameters`) is always querystring/cookie if present or defaults if not, while the default filter definition does not change unless you manually implement something to change `context.IsDefaultReplaced`. This fixes a bunch of weird "show default" bugs as well, since we can now be sure that the default filter is the compiled SetGriddlyDefault(s) (again, unless you've implemented IsDefaultReplaced).

Reply:
Here's a quick test to show that when a default filter is removed, the cookie value (null) is correctly applied after navigating away and return
![chrome-capture-2024-0-30](https://github.com/programcsharp/griddly/assets/3342620/6bbc4043-f61e-499b-8073-b97eec459ba4)
ing:

Reply:
I don't think it's working properly. The Filter Defaults test page no longer applies the defaults:
![image](https://github.com/programcsharp/griddly/assets/8271114/a66cc75a-47b5-43a9-8dbf-679686b666de)
It used to look like this:
![image](https://github.com/programcsharp/griddly/assets/8271114/32dbaf5e-f748-4bc2-b65e-9cc85cd55199)

Reply:
It's working for me in both framework and core examples... 
<img width="682" alt="image" src="https://github.com/programcsharp/griddly/assets/3342620/deafea1a-e915-45d4-b71c-1a5da0433ac5">

Reply:
@ithielnor Re-tested and it seems to be working properly for the most part in .NET Core. The Filter Defaults page is sticky, which probably is what had confused me before.

There _is_ one discrepancy between the .NET 4.x version and the .NET Core version on the "Filter Defaults Example" page.  In the .NET Core version, if you click the "Show All" button, the status shows:
![image](https://github.com/user-attachments/assets/d9f33ec5-8fc8-44bf-829e-f1267f7b6fa8)
If you then refresh the page, it shows:
![image](https://github.com/user-attachments/assets/b6117874-3db7-462e-8c6b-8b11ac1e385f)
In the .NET 4.x version, it continues showing this after the refresh:
![image](https://github.com/user-attachments/assets/d9f33ec5-8fc8-44bf-829e-f1267f7b6fa8)

I think there are some things missing on the .NET Core project. You've made some changes to views in Griddly but not Griddly.NetCore.

Here are some examples:
https://github.com/programcsharp/griddly/pull/121/commits/5e204e2087394ace1f296ed66d67550193feeff2#diff-6ef3205a21ad4c9225aa22537ec82c939e30becd312a7159d06616fda919de9c
https://github.com/programcsharp/griddly/pull/121/commits/f96c524375b5e0ce437f38b9dca294aafa45daa4#diff-6ef3205a21ad4c9225aa22537ec82c939e30becd312a7159d06616fda919de9c